cmd/internal/database/valkey/valkey.go

```diff
@@ -0,0 +1,362 @@
+					podName := os.Getenv("POD_NAME")
+					ordinal := extractOrdinalFromPodName(podName)
+
+					if ordinal == -1 {
+						return fmt.Errorf("failed to extract pod ordinal from POD_NAME: %s", podName)
+					}
+
+					if ordinal != 0 {
+						db.log.Info("elected as backup leader but not pod-0 (won't be Valkey master), skipping restore",
+							"ordinal", ordinal,
+							"electionTime", elapsed.String())
+						return nil
+					}
+
+					db.log.Info("elected as backup leader AND pod-0 (will be Valkey master), performing restore from backup",
+						"ordinal", ordinal,
+						"electionTime", elapsed.String())
+					return db.performRestore(dump)
+				}
+				// Still waiting for election log progress every 10 seconds
+				elapsed := time.Since(startTime)
+				if int(elapsed.Seconds())%10 == 0 && elapsed.Seconds() > 0 {
+					db.log.Debug("waiting for leader election",
+						"elapsed", elapsed.String(),
+						"timeout", "90s")
+				}
+			}
+		}
+	}
+	return db.performRestore(dump)
+}
+
+func (db *Valkey) performRestore(dump string) error {
+	if err := utils.RemoveContents(db.datadir); err != nil {
+		return fmt.Errorf("could not clean database data directory: %w", err)
+	}
+
+	err := utils.Copy(afero.NewOsFs(), dump, path.Join(db.datadir, valkeyDumpFile))
+	if err != nil {
+		return fmt.Errorf("unable to recover: %w", err)
+	}
+	db.log.Info("successfully restored valkey database")
+	return nil
+}
+
+func (db *Valkey) Upgrade(context.Context) error {
+	return nil
+}
+
+func (db *Valkey) Backup(ctx context.Context) error {
+	if !db.clusterMode {
+		isMaster, err := db.isMaster(ctx)
+		if err != nil {
+			return err
+		}
+		if !isMaster {
+			db.log.Info("this database is not master, not taking a backup")
+			return nil
+		}
+	}
+
+	if err := os.RemoveAll(constants.BackupDir); err != nil {
+		return fmt.Errorf("could not clean backup directory: %w", err)
+	}
+
+	if err := os.MkdirAll(constants.BackupDir, 0777); err != nil {
+		return fmt.Errorf("could not create backup directory: %w", err)
+	}
+
+	start := time.Now()
+
+	_, err := db.client.Save(ctx).Result()
+	if err != nil {
+		return fmt.Errorf("could not create a dump: %w", err)
+	}
+
+	dumpFile := path.Join(db.datadir, valkeyDumpFile)
+	db.log.Info("dump created successfully", "file", dumpFile, "duration", time.Since(start).String())
+
+	err = utils.Copy(afero.NewOsFs(), dumpFile, path.Join(constants.BackupDir, valkeyDumpFile))
+	if err != nil {
+		return fmt.Errorf("unable to copy dumpfile to backupdir: %w", err)
+	}
+
+	err = os.Remove(dumpFile)
+	if err != nil {
+		return fmt.Errorf("unable to clean up dump: %w", err)
+	}
+
+	db.log.Debug("successfully took backup of valkey")
+	return nil
+}
+
+func New(
+	log *slog.Logger,
+	datadir string,
+	password *string,
+	statefulSetName string,
+	clusterMode bool,
+	clusterSize int) (*Valkey, error) {
+	v := &Valkey{
+		log:             log,
+		datadir:         datadir,
+		password:        getPassword(password),
```

Comment:
You can replace the `getPassword` function in favor of a function from [metal-lib](https://github.com/metal-stack/metal-lib/blob/master/pkg/pointer/pointer.go):

```suggestion
		password:         pointer.SafeDerefOrDefault(password, ""),
```